### PR TITLE
Add IPv6 GRE support

### PR DESCRIPTION
**Pull request by simon-batardiere**

### CODE REVIEW

**Comment by damjan-we** [comment]
There is one unnecessary merge commit in this PR.
Please rebase and force push the branch.

**Comment by zhichengqiu** [comment]
The code looks good to me

File: src/lib/inet/src/linux/inet_gretap.c [hunk 1]
```diff
@@ -197,8 +243,10 @@ bool inet_gretap_ip4tunnel_set(
 bool inet_gretap_interface_start(inet_gretap_t *self, bool enable)
 {
     int status;
-    char slocal_addr[C_IP4ADDR_LEN];
-    char sremote_addr[C_IP4ADDR_LEN];
+    char slocal_addr[C_IPV6ADDR_LEN];
+    char sremote_addr[C_IPV6ADDR_LEN];
+    char family[32];
+    char type[32];
```
Comment (lkristan-plume) [inline src/lib/inet/src/linux/inet_gretap.c:249]:
```suggestion
    const char *family;
    const char *type;
```

File: src/lib/inet/src/linux/inet_gretap.c [hunk 2]
```diff
@@ -208,26 +256,58 @@ bool inet_gretap_interface_start(inet_gretap_t *self, bool enable)
...
+            if (osn_ip_addr_cmp(&self->in_remote_addr, &OSN_IP_ADDR_INIT) == 0)
+            {
+                LOG(INFO, "inet_gretap: %s: No remote address was specified.", self->inet.in_ifname);
+                return false;
+            }
+
+            snprintf(slocal_addr, sizeof(slocal_addr), PRI_osn_ip_addr, FMT_osn_ip_addr(self->in_local_addr));
+            snprintf(sremote_addr, sizeof(sremote_addr), PRI_osn_ip_addr, FMT_osn_ip_addr(self->in_remote_addr));
+            strncpy(family, "-4", sizeof(family));
+            strncpy(type, "gretap", sizeof(type));
```
Comment (lkristan-plume) [inline src/lib/inet/src/linux/inet_gretap.c:276]:
```suggestion
            family = "-4";
            type = "gretap";
```

strncpy() is an unsafe function (indeed it is safe here, but then again, so would be strcpy)
See: https://www.opensync.io/s/EDE-020-041-501_OpenSync_Coding_Standard.pdf

File: src/lib/inet/src/linux/inet_gretap.c [hunk 3]
```diff
@@ -208,26 +256,58 @@ bool inet_gretap_interface_start(inet_gretap_t *self, bool enable)
...
+            if (osn_ip6_addr_cmp(&self->in6_remote_addr, &OSN_IP6_ADDR_INIT) == 0)
+            {
+                LOG(INFO, "inet_gretap: %s: No remote address was specified.", self->inet.in_ifname);
+                return false;
+            }
+
+            snprintf(slocal_addr, sizeof(slocal_addr), PRI_osn_ip6_addr, FMT_osn_ip6_addr(self->in6_local_addr));
+            snprintf(sremote_addr, sizeof(sremote_addr), PRI_osn_ip6_addr, FMT_osn_ip6_addr(self->in6_remote_addr));
+            strncpy(family, "-6", sizeof(family));
+            strncpy(type, "ip6gretap", sizeof(type));
```
Comment (lkristan-plume) [inline src/lib/inet/src/linux/inet_gretap.c:295]:
```suggestion
            family = "-6";
            type = "ip6gretap";
```

File: src/lib/inet/src/linux/inet_gretap.c [hunk 4]
```diff
@@ -208,26 +256,58 @@ bool inet_gretap_interface_start(inet_gretap_t *self, bool enable)
...
+            break;
 
-        if (osn_ip_addr_cmp(&self->in_remote_addr, &OSN_IP_ADDR_INIT) == 0)
-        {
-            LOG(INFO, "inet_gretap: %s: No remote address was specified.", self->inet.in_ifname);
+        default:
+            LOG(INFO, "inet_gretap: invalid address family.");
```
Comment (lkristan-plume) [inline src/lib/inet/src/linux/inet_gretap.c:299]:
```suggestion
            LOG(ERR, "inet_gretap: invalid address family.");
```

File: src/lib/inet/src/linux/inet_gretap.c [hunk 5]
```diff
@@ -208,26 +256,58 @@ bool inet_gretap_interface_start(inet_gretap_t *self, bool enable)
...
             return false;
+            break;
```
Comment (lkristan-plume) [inline src/lib/inet/src/linux/inet_gretap.c:301]:
```suggestion
```

Unreachable statement.

File: src/nm2/inc/nm2_iface.h [hunk 6]
```diff
@@ -87,6 +87,7 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
     M(NM2_IFTYPE_VLAN,      "vlan")     \
     M(NM2_IFTYPE_BRIDGE,    "bridge")   \
     M(NM2_IFTYPE_GRE,       "gre")      \
+    M(NM2_IFTYPE_GRE6,      "gre6")      \
```
[the remark this answers is not in the document]
Comment (simon-batardiere) [reply src/nm2/inc/nm2_iface.h]: OK

File: src/lib/inet/inc/inet_gretap.h [hunk 7]
```diff
@@ -38,14 +38,21 @@ struct __inet_gretap
     /* Subclass from inet_eth_t, include base and inet for convenience */
     union
     {
-        inet_t      inet;
-        inet_base_t base;
-        inet_eth_t  eth;
+        inet_t          inet;
+        inet_base_t     base;
+        inet_eth_t      eth;
```
[the remark this answers is not in the document]
Comment (simon-batardiere) [reply src/lib/inet/inc/inet_gretap.h]: OK

File: src/lib/inet/inc/inet_gretap.h [hunk 8]
```diff
@@ -38,14 +38,21 @@ struct __inet_gretap
...
     };
 
-    char            in_ifparent[C_IFNAME_LEN];      /* Parent interface */
-    osn_ip_addr_t   in_local_addr;                  /* Local IPv4 address */
-    osn_ip_addr_t   in_remote_addr;                 /* Remote IPv4 address */
+    char                in_ifparent[C_IFNAME_LEN];      /* Parent interface */
+    int                 family;                         /* Family */
```
[the remark this answers is not in the document]
Comment (simon-batardiere) [reply src/lib/inet/inc/inet_gretap.h]: OK

File: src/lib/inet/inc/inet_gretap.h [hunk 9]
```diff
@@ -38,14 +38,21 @@ struct __inet_gretap
...
+    union {
+        osn_ip_addr_t   in_local_addr;                  /* Local IPv4 address */
+        osn_ip6_addr_t  in6_local_addr;                 /* Local IPv6 address */
+    };
+    union {
+        osn_ip_addr_t   in_remote_addr;                 /* Remote IPv4 address */
+        osn_ip6_addr_t  in6_remote_addr;                /* Remote IPv6 address */
```
[the remark this answers is not in the document]
Comment (simon-batardiere) [reply src/lib/inet/inc/inet_gretap.h]: OK

File: src/lib/inet/src/linux/inet_gretap.c (hunk 2 above)
Comment (simon-batardiere) [reply src/lib/inet/src/linux/inet_gretap.c]: OK

File: src/lib/inet/src/linux/inet_gretap.c (hunk 4 above)
Comment (simon-batardiere) [reply src/lib/inet/src/linux/inet_gretap.c]: OK

File: src/lib/inet/src/linux/inet_gretap.c (hunk 5 above)
Comment (simon-batardiere) [reply src/lib/inet/src/linux/inet_gretap.c]: OK

File: src/lib/inet/src/linux/inet_gretap.c (hunk 1 above)
Comment (simon-batardiere) [reply src/lib/inet/src/linux/inet_gretap.c]: OK

File: src/lib/inet/src/linux/inet_gretap.c (hunk 3 above)
Comment (simon-batardiere) [reply src/lib/inet/src/linux/inet_gretap.c]: OK

**Comment by simon-batardiere** [comment]
@mitja-plume Before to start to change the implementation, I would like to clarify your request:
1- Could we keep just one file?
2- Do I have to duplicate all the code, or just the function `inet_gre_new`. And the the new function `inet_gre6_new` will also call `inet_gretap_new`.

**Comment by mitja-plume** [comment]
@simon-batardiere 

> 1- Could we keep just one file?

Yes.

> 2- Do I have to duplicate all the code, or just the function inet_gre_new. And the the new function inet_gre6_new will also call inet_gretap_new.

You need to duplicate just the function `inet_gre_new` and create `inet_gre6_new()`. I would still avoid calling `inet_gretap_new()` in this case, but I think it break anything so I'm OK with it.

**Comment by dejan-pojbic-plume** [comment]
Closing since GRE IPv6 support was added in newer releases.

**Pull request closed without merging**